src/terrain/terrain.js

```diff
@@ -399,9 +399,8 @@ export class Terrain extends Elevation {
             this._initializing = tr._centerAltitude === 0 && this.getAtPoint(MercatorCoordinate.fromLngLat(tr.center), -1) === -1;
             this._emptyDEMTextureDirty = !this._initializing;
         }
-
         const options = this.painter.options;
-        this.renderCached = (options.zooming || options.moving || options.rotating || !!this.forceRenderCached) && !this._invalidateRenderCache;
+        this.renderCached = (options.zooming || options.moving || options.rotating || !!this.forceRenderCached) && !this._invalidateRenderCache && !options.styleDirty;
```

Comment:
There's similar implementation used for crossfade below:
https://github.com/mapbox/mapbox-gl-js/blob/5db846990041e3f3b7c09afc923058fd7eddee01/src/terrain/terrain.js#L830-L836

should we add this.style.hasTransitions() there instead?

Reply:
Yes I agree, I also think that would be a better place to consolidate this logic.

Reply:
Agreed. I've updated the PR by moving the introduced logic to `_shouldDisableRenderCache`. I was initially concerned if some of the transition events could be missed because `style.update()` was called before `painter.render()` leading to "transition" flags being clear before rendering of the frame was done. But this works correctly as render cache contents gets updated in the very first frame after transitions are finished